src/Elmish.WPF/WpfProgram.fs

```diff
@@ -0,0 +1,133 @@
+﻿namespace Elmish.WPF
+
+open System.Windows
+open Microsoft.Extensions.Logging
+open Microsoft.Extensions.Logging.Abstractions
+open Elmish
+
+
+type WpfProgram<'model, 'msg> =
+  internal {
+    ElmishProgram: Program<unit, 'model, 'msg, Binding<'model, 'msg> list>
+    LoggerFactory: ILoggerFactory
+    /// Only log calls that take at least this many milliseconds. Default 1.
+    PerformanceLogThreshold: int
+  }
+
+
+[<RequireQualifiedAccess>]
+module WpfProgram =
+
+
+  let private fromElmishProgram program =
+    { ElmishProgram = program
+      LoggerFactory = NullLoggerFactory.Instance
+      PerformanceLogThreshold = 1 }
+
+
+  /// Creates a WpfProgram that does not use commands.
+  let mkSimple
+      (init: unit -> 'model)
+      (update: 'msg  -> 'model -> 'model)
+      (bindings: unit -> Binding<'model, 'msg> list) =
+    Program.mkSimple init update (fun _ _ -> bindings ())
+    |> fromElmishProgram
+
+
+  /// Creates a WpfProgram that uses commands
+  let mkProgram
+      (init: unit -> 'model * Cmd<'msg>)
+      (update: 'msg  -> 'model -> 'model * Cmd<'msg>)
+      (bindings: unit -> Binding<'model, 'msg> list) =
+    Program.mkProgram init update (fun _ _ -> bindings ())
+    |> fromElmishProgram
+
+
+  /// Starts the Elmish dispatch loop, setting the bindings as the DataContext
+  /// for the specified FrameworkElement. Non-blocking. This is a low-level function;
+  /// for normal usage, see runWindow and runWindowWithConfig.
+  let startElmishLoop
+      (element: FrameworkElement)
+      (program:  WpfProgram<'model, 'msg>) =
+    let mutable lastModel = None
+
+    let setState model dispatch =
+      match lastModel with
+      | None ->
+          let bindings = Program.view program.ElmishProgram model dispatch
+          let vm = ViewModel<'model, 'msg>(model, dispatch, bindings, program.PerformanceLogThreshold, "main", program.LoggerFactory)
+          element.DataContext <- vm
+          lastModel <- Some vm
+      | Some vm ->
+          vm.UpdateModel model
+
+    let uiDispatch (innerDispatch: Dispatch<'msg>) : Dispatch<'msg> =
+      fun msg -> element.Dispatcher.Invoke(fun () -> innerDispatch msg)
+
+    let logMsgAndModel (msg: 'msg) (model: 'model) = 
+      program.LoggerFactory.CreateLogger("Elmish.WPF.Update").LogTrace("New message: {Message}", msg)
+      program.LoggerFactory.CreateLogger("Elmish.WPF.Update").LogTrace("Updated state:\n{Model}", model)
+
+    let logError (msg: string, ex: exn) =
+      program.LoggerFactory.CreateLogger("Elmish.WPF.Update").LogError(ex, msg)
```

Comment:
Can you optimize this by first creating the logger and then returning a function that uses it?

Reply:
I tried, but the logs didn't show up. I think there may be a race condition. In any case, AFAIK `CreateLogger` is very cheap and intended to be called frequently. I would guess it's cached at the framework level for most log frameworks. Let me know if this is incorrect.

Reply:
Hm, seems to work now. I'll make the change. Perhaps I'm mixing this with another thing I came across earlier.

Reply:
Yep, definitely mixing things (I know what I mixed it up with now). Pushed a change where all loggers are created only once.